File: packages/capacitor/ios/Sources/KeywardPlugin/KeywardPlugin.swift
import Foundation
import Capacitor
#if canImport(Keyward)
import Keyward
#endif

@objc(KeywardPlugin)
public class KeywardPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "KeywardPlugin"
    public let jsName = "Keyward"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "get", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "set", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "remove", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "keys", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clear", returnType: CAPPluginReturnPromise),
    ]

    @objc func get(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Missing required parameter: key")
            return
        }
        let value = Keyward.getRaw(key: key)
        call.resolve(["value": value as Any])
    }

    @objc func set(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Missing required parameter: key")
            return
        }
        guard let value = call.getString("value") else {
            call.reject("Missing required parameter: value")
            return
        }
        do {
            try Keyward.setRaw(key: key, value: value)
            call.resolve()
        } catch {
            call.reject("Keyward: setRaw failed", nil, error)
        }
    }

    @objc func remove(_ call: CAPPluginCall) {
        guard let key = call.getString("key") else {
            call.reject("Missing required parameter: key")
            return
        }
        Keyward.removeRaw(key: key)
        call.resolve()
    }

    @objc func keys(_ call: CAPPluginCall) {
        let allKeys = Keyward.keys()
        call.resolve(["keys": allKeys])
    }

    @objc func clear(_ call: CAPPluginCall) {
        Keyward.clearAll()
        call.resolve()
    }
}

File: packages/capacitor/android/src/main/java/org/keyward/capacitor/KeywardPlugin.java
package org.keyward.capacitor;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.keyward.Keyward;
import org.keyward.KeywardException;

import java.util.List;

@CapacitorPlugin(name = "Keyward")
public class KeywardPlugin extends Plugin {

    @Override
    public void load() {
        try {
            Keyward.setBackend(new EncryptedPrefsBackend(getContext()));
        } catch (Exception e) {
            android.util.Log.e("KeywardPlugin", "Failed to initialize backend", e);
        }
    }

    @PluginMethod
    public void get(PluginCall call) {
        String key = call.getString("key");
        if (key == null) {
            call.reject("Missing required parameter: key");
            return;
        }
        String value = Keyward.getRaw(key);
        JSObject result = new JSObject();
        result.put("value", value);
        call.resolve(result);
    }

    @PluginMethod
    public void set(PluginCall call) {
        String key = call.getString("key");
        if (key == null) {
            call.reject("Missing required parameter: key");
            return;
        }
        String value = call.getString("value");
        if (value == null) {
            call.reject("Missing required parameter: value");
            return;
        }
        try {
            Keyward.setRaw(key, value);
            call.resolve();
        } catch (KeywardException e) {
            call.reject("Keyward: setRaw failed", e);
        }
    }

    @PluginMethod
    public void remove(PluginCall call) {
        String key = call.getString("key");
        if (key == null) {
            call.reject("Missing required parameter: key");
            return;
        }
        Keyward.removeRaw(key);
        call.resolve();
    }

    @PluginMethod
    public void keys(PluginCall call) {
        List<String> allKeys = Keyward.keys();
        JSObject result = new JSObject();
        result.put("keys", new JSArray(allKeys));
        call.resolve(result);
    }

    @PluginMethod
    public void clear(PluginCall call) {
        Keyward.clearAll();
        call.resolve();
    }
}
